### PR TITLE
Extract cleanup changes from #21793.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Extract `stubgen.h` and `stubgen.cpp` changes from #21793. Add a comment block explaining why we marshal arrays the way we do so someone doesn't try to do #21793 again.

cc: @jkotas @AaronRobinsonMSFT

### CODE REVIEW

File: src/vm/ilmarshalers.cpp [hunk 1]
```diff
@@ -4686,7 +4690,8 @@ void ILHiddenLengthArrayMarshaler::EmitMarshalArgumentCLRToNative()
         // native = pinnedLocal + dataOffset
         m_pcsMarshal->EmitLDLOC(dwPinnedLocal);
         m_pcsMarshal->EmitCONV_I();
-        m_pcsMarshal->EmitLDC(m_pargs->na.m_optionalbaseoffset);
+        m_pcsMarshal->EmitLDLOC(dwPinnedLocal);
+        m_pcsMarshal->EmitCALL(METHOD__ARRAY__GET_DATA_PTR_OFFSET_INTERNAL, 1, 1);
```
Comment (jkotas) [inline src/vm/ilmarshalers.cpp]: Is this going to make array marshaling slower?

File: src/vm/ilmarshalers.cpp [hunk 2]
```diff
@@ -3804,11 +3800,18 @@ void ILNativeArrayMarshaler::EmitMarshalArgumentCLRToNative()
...
+        // Additionally, we need to ensure that we do not pass non-null for a zero-length
+        // array when interacting with GDI/GDI+ since they fail on null arrays but succeed
+        // on 0-length arrays.
         EmitLoadManagedValue(m_pcsMarshal);
         m_pcsMarshal->EmitSTLOC(dwPinnedLocal);
         m_pcsMarshal->EmitLDLOC(dwPinnedLocal);
         m_pcsMarshal->EmitCONV_I();
-        m_pcsMarshal->EmitLDC(m_pargs->na.m_optionalbaseoffset);
+        m_pcsMarshal->EmitLDLOC(dwPinnedLocal);
+        m_pcsMarshal->EmitCALL(METHOD__ARRAY__GET_DATA_PTR_OFFSET_INTERNAL, 1, 1);
```
Comment (jkotas) [inline src/vm/ilmarshalers.cpp]: Is this call going to make array marshaling slower?

Comment (jkoritzinsky) [reply src/vm/ilmarshalers.cpp]: I don't think so, but I might be wrong. I'll go revert this part of the change.

Comment (jkotas) [reply src/vm/ilmarshalers.cpp]: I think you can compute the offset right here without passing it around.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Windows_NT x64 Checked Innerloop Build and Test
@dotnet-bot test Windows_NT x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0)
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test (Jit - TieredCompilation=0)
@dotnet-bot test Windows_NT x86 Release Innerloop Build and Test

**Pull request merged**